editor/src/components/editor/store/collaborative-editing.ts

```diff
@@ -609,3 +612,119 @@ function synchroniseParseSuccessToCollabFile(
     ImportDetailsKeepDeepEquality,
   )
 }
+
+export interface ClaimProjectOwnership {
+  type: 'CLAIM_PROJECT_OWNERSHIP'
+  projectID: string
+  collaborationEditor: string
+}
+
+export function claimProjectOwnership(
+  projectID: string,
+  collaborationEditor: string,
+): ClaimProjectOwnership {
+  return {
+    type: 'CLAIM_PROJECT_OWNERSHIP',
+    projectID: projectID,
+    collaborationEditor: collaborationEditor,
+  }
+}
+
+export interface ClearAllOfCollaboratorsOwnership {
+  type: 'CLEAR_ALL_OF_COLLABORATORS_OWNERSHIP'
+  collaborationEditor: string
+}
+
+export function clearAllOfCollaboratorsOwnership(
+  collaborationEditor: string,
+): ClearAllOfCollaboratorsOwnership {
+  return {
+    type: 'CLEAR_ALL_OF_COLLABORATORS_OWNERSHIP',
+    collaborationEditor: collaborationEditor,
+  }
+}
+
+export type CollaborationRequest = ClaimProjectOwnership | ClearAllOfCollaboratorsOwnership
+
+export interface ClaimProjectOwnershipResult {
+  type: 'CLAIM_PROJECT_OWNERSHIP_RESULT'
+  successfullyClaimed: boolean
+}
+
+export interface ClearAllOfCollaboratorsOwnershipResult {
+  type: 'CLEAR_ALL_OF_COLLABORATORS_OWNERSHIP_RESULT'
+}
+
+export type CollaborationResponse =
+  | ClaimProjectOwnershipResult
+  | ClearAllOfCollaboratorsOwnershipResult
+
+const collaborationEditor = UUID()
+
+async function callCollaborationEndpoint(
+  request: CollaborationRequest,
+): Promise<CollaborationResponse> {
+  const url = `${UTOPIA_BACKEND}collaboration`
+  const response = await fetch(url, {
+    method: 'PUT',
+    credentials: 'include',
+    headers: HEADERS,
+    mode: MODE,
+    body: JSON.stringify(request),
+  })
+  if (response.ok) {
+    return response.json()
+  } else {
+    throw new Error(`server responded with ${response.status} ${response.statusText}`)
+  }
+}
+
+export async function claimControlOverProject(projectID: string | null): Promise<boolean | null> {
+  if (projectID == null || !isFeatureEnabled('Baton Passing For Control')) {
+    return null
+  }
+
+  const request = claimProjectOwnership(projectID, collaborationEditor)
+  const response = await callCollaborationEndpoint(request)
+  if (response.type === 'CLAIM_PROJECT_OWNERSHIP_RESULT') {
+    return response.successfullyClaimed
+  } else {
+    throw new Error(`Unexpected response: ${JSON.stringify(response)}`)
+  }
+}
+
+export async function releaseControl(): Promise<void> {
+  if (isFeatureEnabled('Baton Passing For Control')) {
+    const request = clearAllOfCollaboratorsOwnership(collaborationEditor)
+
+    const response = await callCollaborationEndpoint(request)
```

Comment:
is this also somehow updating the porjectServerState?

Reply:
It's not, but only by virtue of it only being called when the editor closes so there's not much point.

Reply:
ah i see! I thought this is also being called if you tab away, or somehow send the editor to the background

Reply:
This much is true, as a result of the switch of event handler.

That complicates things because now it needs access to the dispatch function so it can update that state, I'm actually tempted to revert that change because it means potentially losing control when the browser just fancies it rather than when the tab is closed.

Reply:
Reverted event handler change in 4267c3e67.